internal/controller/services/gateway/gateway_controller_actions.go

```diff
@@ -0,0 +1,201 @@
+		"gateway", "odh-gateway",
+		"namespace", gatewayInstance.Spec.Namespace,
+		"domain", domain,
+		"certificateType", getCertificateType(gatewayInstance))
+
+	return nil
+}
+
+func createGatewayClass(rr *odhtypes.ReconciliationRequest) error {
+	gatewayClass := &gwapiv1.GatewayClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "odh-gateway-class",
+			Labels: map[string]string{
+				"app.kubernetes.io/managed-by": "opendatahub-operator",
+				"opendatahub.io/internal":      "true",
+			},
+		},
+		Spec: gwapiv1.GatewayClassSpec{
+			ControllerName: "openshift.io/gateway-controller/v1",
+		},
+	}
+
+	return rr.AddResources(gatewayClass)
+}
+
+func handleCertificates(ctx context.Context, rr *odhtypes.ReconciliationRequest, gatewayInstance *serviceApi.GatewayConfig, domain string) (string, error) {
+	certConfig := gatewayInstance.Spec.Certificate
+	if certConfig == nil {
+		certConfig = &infrav1.CertificateSpec{
+			Type: infrav1.OpenshiftDefaultIngress,
+		}
+	}
+
+	secretName := certConfig.SecretName
+	if secretName == "" {
+		secretName = fmt.Sprintf("%s-tls", gatewayInstance.Name)
+	}
+
+	switch certConfig.Type {
+	case infrav1.OpenshiftDefaultIngress:
+		return handleOpenshiftDefaultCertificate(ctx, rr, gatewayInstance, secretName)
+	case infrav1.SelfSigned:
+		return handleSelfSignedCertificate(ctx, rr, gatewayInstance, secretName, domain)
+	case infrav1.Provided:
+		return secretName, nil
+	default:
+		return "", fmt.Errorf("unsupported certificate type: %s", certConfig.Type)
+	}
+}
+
+func handleOpenshiftDefaultCertificate(ctx context.Context, rr *odhtypes.ReconciliationRequest, gatewayInstance *serviceApi.GatewayConfig, secretName string) (string, error) {
+	err := cluster.PropagateDefaultIngressCertificate(ctx, rr.Client, secretName, gatewayInstance.Spec.Namespace)
+	if err != nil {
+		return "", fmt.Errorf("failed to propagate default ingress certificate: %w", err)
+	}
+
+	return secretName, nil
+}
+
+func handleSelfSignedCertificate(ctx context.Context, rr *odhtypes.ReconciliationRequest,
+	gatewayInstance *serviceApi.GatewayConfig, secretName string, domain string) (string, error) {
+	hostname := "odh-gateway." + domain
+
+	err := cluster.CreateSelfSignedCertificate(
+		ctx,
+		rr.Client,
+		secretName,
+		hostname,
+		gatewayInstance.Spec.Namespace,
+	)
+	if err != nil {
+		return "", fmt.Errorf("failed to create self-signed certificate: %w", err)
+	}
+
+	return secretName, nil
+}
+
+func getCertificateType(gatewayInstance *serviceApi.GatewayConfig) string {
+	if gatewayInstance.Spec.Certificate == nil {
+		return string(infrav1.OpenshiftDefaultIngress)
+	}
+	return string(gatewayInstance.Spec.Certificate.Type)
+}
+
+func createGateway(rr *odhtypes.ReconciliationRequest, gatewayInstance *serviceApi.GatewayConfig, certSecretName string, domain string) error {
+	listeners := createListeners(certSecretName, domain)
+
+	gateway := &gwapiv1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "odh-gateway",
```

Comment:
Suggest putting this name in a constant, since it will be referenced  in multiple places, including as parent when creating attached routes.  
Also since this name is going to be user visible should it be "rhoai"  instead of "odh", or at least add configuration to name it different downstream?